File: aoc/day2/day2.ml
open Core

let split_lines lines =
  List.map lines ~f:(fun line ->
      String.split line ~on:' ' |> List.map ~f:int_of_string)

let safe_report report =
  let allowable = [ 1; 2; 3 ] in
  let rec check_direction direction i =
    if i >= List.length report then true
    else
      let diff = List.nth_exn report i - List.nth_exn report (i - 1) in
      if not (List.mem allowable (abs diff) ~equal:Int.equal) then false
      else
        match direction with
        | None ->
            let new_direction =
              if diff > 0 then Some "increase" else Some "decrease"
            in
            check_direction new_direction (i + 1)
        | Some "increase" ->
            if diff <= 0 then false else check_direction direction (i + 1)
        | Some "decrease" ->
            if diff >= 0 then false else check_direction direction (i + 1)
        | _ -> false
  in
  check_direction None 1

let puzzle1 reports =
  List.fold_left reports ~init:0 ~f:(fun acc report ->
      if safe_report report then acc + 1 else acc)

let puzzle2 reports =
  List.fold_left reports ~init:0 ~f:(fun acc report ->
      if safe_report report then acc + 1
      else
        let rec try_remove_element i =
          if i >= List.length report then false
          else
            let modified_report = List.filteri report ~f:(fun j _ -> j <> i) in
            if safe_report modified_report then true
            else try_remove_element (i + 1)
        in
        if try_remove_element 0 then acc + 1 else acc)

let solve filename =
  filename |> Read_file.read_file |> split_lines |> fun parsed_input ->
  let res1 = puzzle1 parsed_input in
  Printf.printf "The Puzzle 1 solution is: %d\n" res1;

  let res2 = puzzle2 parsed_input in
  Printf.printf "The Puzzle 2 solution is: %d\n" res2

File: aoc/day2/day2.py
input = []
with open("input.txt", "r") as file:
    for line in file:
        report = list(map(int,line.split()))
        input.append(report)

allowable = [1,2,3]
res = 0

def safe(report):
    direction = None
    for j in range(1, len(report)):
        diff = report[j] - report[j - 1]
        if abs(diff) not in allowable:
            return False
        if direction is None:
            direction = "increase" if diff > 0 else "decrease"
        elif (direction == "increase" and diff <= 0) or (direction == "decrease" and diff >= 0):
            return False
    return True

res1 = 0
for report in input:
    if safe(report):
        res1 += 1

res2 = 0
for report in input:
    if safe(report):
        res2 += 1
    else:
        for i in range(len(report)):
            modified_report = report[:i] + report[i+1:]
            if safe(modified_report):
                res2 += 1
                break

print(f'The Puzzle 1 solution is: {res1}')
print(f'The Puzzle 2 solution is: {res2}')